### ISSUE TITLE
Authorization based on a action paramter value.  Identity 3 docs lacking.

### ISSUE THREAD

**Issue by mikes-gh** [aspnet/Mvc#3757]
I have done a lot of research but am still not sure if I am doing this correctly.
The docs are still very light on identity 3 and recommendations and examples seem contradictory or dated to earlier Identity

Given an ApplicationUser class extended to include a list of Authorized account numbers I want to restrict the user to only view statements (and other actions based) on their authorized accounts).  I would think this is a very common design however most of the articles on the net refer to previous versions of identity.

(PS I am injecting UserManager in the Controller constructor)

Here is my action
```csharp
public IActionResult GetStatement(int accountNo,DateTime startDate,DateTime endDate)
{
	var user = userManager.Users
		.Include(u => u.AuthorisedAccounts)
		.Where(u => u.Id == User.GetUserId())
		.FirstOrDefault();
	if (user.AuthorisedAccounts != null)
	{
		foreach (var account in user.AuthorisedAccounts)
		{
			if (account.AccountNo == accountNo)
				return View(statementService.GetStatement(accountNo, startDate, endDate, 0));
		}
	}
	return HttpUnauthorized();
}
```
I cant help feeling there is a better way?
Basically I want to authorize based on the action parameter."accountNo"

**Comment by tuespetre** [comment]
Hey @mikes-gh, check out the [fenced code blocks syntax for GitHub-flavored MarkDown.](https://help.github.com/articles/github-flavored-markdown/#fenced-code-blocks) It will make your code examples 'nice and tidy'.

As to your question, I would recommend checking out this article about using policies for authorization: http://leastprivilege.com/2015/10/12/the-state-of-security-in-asp-net-5-and-mvc-6-authorization/

By defining this as a policy, you can reuse the policy across multiple actions without having to redefine the logic every time. Now, there are some downsides, namely that you will need to ensure that your route or action parameters always use the same name where you use this policy (perhaps among the ASP.NET community we can dream up a solution for that, though!)

**Comment by mikes-gh** [comment]
I did read through that's.  Still not sure how to proceed.  .  What type of policy would I use.  Resource based doesn't seem quite right?   I think I maybe should but this on stack overflow as it is probably my lack of knowledge.

**Comment by mikes-gh** [comment]
Thanks for your help but I need a bit more hand holding.  Moved to

http://stackoverflow.com/questions/34293133/authorisation-policy-in-identity-3-mvc-6

Please close.

**Comment by tuespetre** [comment]
You can close the issue on your own. I will hop over to StackOverflow and continue discussing with you.

**Comment by blowdart** [comment]
In this case you'd use resource based, with the account being the resource. The documentation for this is at https://docs.asp.net/en/latest/security/authorization/resourcebased.html

To start with you'd define an operation of Read, 

```c#
public static class Operations
{
    public static OperationAuthorizationRequirement Read =
        new OperationAuthorizationRequirement   { Name = "Read" };
}
```

Now you'd have a policy for AccountAccess

```c#
public class AccountAuthorizationHandler : AuthorizationHandler<
    OperationAuthorizationRequirement, Account>
{
    IUserManager _userManager;

    public AccountAuthorizationHandler(IUserManager userManager)
    {
        _userManager = userManager;
    }

    protected override void Handle(AuthorizationContext context,
                                   OperationAuthorizationRequirement requirement,
                                   Account resource)
    {
        // Pull the user ID claim out from the context.User
        var userId = context.User.....
        // Get the current user's account numbers.       
        var user = userManager.Users
            .Include(u => u.AuthorisedAccounts)
            .Where(u => u.Id == userId)
            .FirstOrDefault();
    }

    // Now check if the user's account numbers match the resource accountNumber, and 
    // also check the operation type, in case you want to vary based on create, view etc.
    if (user.AuthorisedAccounts.Contains(resource.AccountId &&
        requirement.Name == "View")
   {
      context.Succeed(requirement);
   } 
}
```

After that register your policy in the DI container, within configure services;

```c#
public void ConfigureServices(IServiceCollection services)
{
    services.AddMvc();

    services.AddAuthorization();

    services.AddInstance<IAuthorizationHandler>(new Authorization.AccountAuthorizationHandler());
}
```

Then, within your controller, after you've loaded the account resource you'd do something like

```c#
public async Task<IActionResult> Edit(int accountId)
{
    Account account = accountManager.Find(accountId);

    if (account == null)
    {
        return new HttpNotFoundResult();
    }

    if (await authorizationService.AuthorizeAsync(User, document, "AccountAccess"))
    {
        return View(account);
    }
    else
    {
        return new ChallengeResult();
    }
}
```

**Comment by blowdart** [comment]
Note that **none** this is part of Identity, it's authorization, which is entirely separate, which may be why you couldn't find documentation, you were looking in the wrong place :)

**Comment by tuespetre** [comment]
Excellent writeup @blowdart, thank you

**Comment by mikes-gh** [comment]
@blowdart finally iget it thank you so much

````
if (await authorizationService.AuthorizeAsync(User, document, "AccountAccess"))
````

Would I be right in assuming

```
if (await authorizationService.AuthorizeAsync(User, account, "View"))
```
Would be closer to the AccountAuthorizationHandler example?

**Comment by blowdart** [comment]
Yea, sorry, messed up sample, edited

**Comment by blowdart** [comment]
I also forgot that in the docs, whoops. I'll get that addressed today.
